### DISCUSSION TITLE
Chat in rooms before joining

### DISCUSSION THREAD

**Discussion by JaminShanti** [jamulussoftware/jamulus#2038] [Ideas]
When Jamming, would be really nice if you could open chats with rooms before connecting.  Maybe limit this but 1 or 2 extra chats would be nice.  Also nice for rooms that are audience rooms, the host can have both chats open for big get togethers.  Really a work around because Jamulas starts to fall apart at 16 clients.

**Comment by dcorson-ticino-com** [comment]
>  Really a work around because Jamulas starts to fall apart at 16 clients.

The user limit depends on the processing power of the server and the quality of the servers' Internet connection.
16 clients seems to be about the limit for Pi based servers.  
Throw better hardware at it or go for a cloud based solution and you should have no problems with many more clients.

**Comment by JaminShanti** [comment]
Chat connections should not count towards the max connections, as they don't require any audio management resources.

**Reply by softins** [reply]
One problem with doing this is that the Jamulus protocol does not currently have a concept of connection distinct from an audio stream. There is no "connect" message; the server detects a new connection just by starting to receive an audio stream from a new IP address and port. This is largely due to the incremental way in which Jamulus evolved over the years from 2006 to now, with a strong emphasis on backward compatibility. So currently it is not possible to join for chat without joining for audio.

A version 2 of the Jamulus protocol could usefully implement the concepts of "join server" and "join audio" separately, but this would be a significant amount of work and be incompatible with the existing userbase.

**Reply by gene96817** [reply]
It is possible to design a version 2 protocol that would fall back to the version 1 operation if the client doesn't follow the version 2 protocol. It does make the server more complicated, but possible.

**Reply by softins** [reply]
That's true, but needs significant effort from someone (or more than one) who has both the time and the motivation! None of us does it as our day job.

**Reply by gene96817** [reply]
I agree. Put this in the "big task lacking an interested developer". 
Personally, I am not very interested in expanding the chat functions. As a participant, the chat is a distraction from jamming. For spectators, their chat, too, would be a distraction. At the most, the spectators can use the chat to announce a separate/separate messaging thread.

**Reply by JaminShanti** [reply]
Solutions over technology.
Getting consensus over a worthwhile problem to solve is difficult with an open source. We all bring our personal experiences and overweight them as affected by all. I personally experience is that rooms are full and I don't know if it's worth waiting. Another one I'm in a okay public room That's almost full. Do I leave it to go check out another room or stay content. Another one to solve, room is at max but several people are listening taking up space. Music is a performance medium so you always need an audience. It would seem this would be worth technically solving.

**Reply by gene96817** [reply]
I agree having an audience is nice. As a virtual performance space, it would be nice to have the ambient sound of an appreciative audience, just like a physical performance space. (A chat window is a good alternative.) The "ambient sound of an appreciative audience" suggests an interaction between performers and audience. Any advancement in this direction would be wonderful.

Related, and different, we could streamcast one audio channel for listeners. Then only one channel is needed for any number of spectators.

**Comment by gilgongo** [comment]
Aside from limiting the number of "unecessary" clients connected to the server, the primary (perhaps only?) advantage of this idea over chat systems like Whatsapp (which are of course routinely used with Jamulus), is that chats are automatically related to the music on the server. You wouldn't have to find and join a third party system. 

One possible step in that direction would be to enable a "chat link" option for the server which showed that link in the directory listing. Anyone wanting to chat would then be able to use that link without joining the session itself. 

Hard to kow why anyone would join such a chat without wanting to also listen to the music though! Perhaps this POC could be modified to  supply audio to a web chat interface for that? https://github.com/jamulussoftware/jamulus/discussions/2039

**Comment by gilgongo** [comment]
Oh BTW:

> Jamulas starts to fall apart at 16 clients.

It depends:


https://user-images.githubusercontent.com/4561747/134595471-34286a99-152e-4870-b9dc-eb545cf81ef6.mp4

**Reply by JaminShanti** [reply]
your post is impressive. What server type, location to connections, average connection speed of clients.  I hear Jitter.  I am not undervaluing this, its quite impressive.  I want to know more about the technical aspect of your example, agreeing it was a musical success.  Unless your going to buy public server to host, you have to assume your aggregate will be much smaller sized free-ish servers.

**Reply by gilgongo** [reply]
I don't actually know about the server spec on this, it was posted someone from the [Deutsche Chorjugend](https://www.youtube.com/channel/UCFmlPFaZzJvxw2qvcDOXSuw) choir a while ago. I'd expect you'd need more resources than the average Jamulus host (500Kb/s per client multiplied by 100 clients is about 50Mbit/s after all), but probably the more challenging thing is making sure all the clients are connecting OK.

**Reply by Andrew999999** [reply]
There's a fuller video here showing their practice.  Recording quality is awful but you get the idea.  https://youtu.be/Z9BJhhi1zEQ

**Comment by mcfnord** [comment]
The welcome message could include a link to a chat community. Anyone can get the welcome message, even without connecting, unless the room is full, of course. Maybe one page could aggregate the welcome messages of all active servers.

**Reply by gilgongo** [reply]
Yes, although I think the OP meant chatting before you joined the server (and/or so that "chatters" didn't take up channels).

> Maybe one page could aggregate the welcome messages of all active servers.

That does in a fact exist if you click on the server names on https://explorer.jamulus.io/

**Reply by pljones** [reply]
That doesn't _aggregate the welcome messages_ and it doesn't work if the room is full.  Hence maybe a page that did that would be useful (i.e. caching them).

**Reply by JaminShanti** [reply]
I've been using https://jamulus.live/. For some of this.
